ecl/hql/hqlgram2.cpp

```diff
@@ -11022,107 +11146,111 @@ void HqlGram::syntaxError(const char *s, int token, int *expected)
     {
-        for (int j = 0; expected[j] ; j++)
+        if (containsToken(UNKNOWN_ID, expected))
         {
-            if (expected[j]==UNKNOWN_ID)
+            msg.append("Identifier '");
+            if (yytext && *yytext)
             {
-                msg.append("Identifier '");
-                if (yytext && *yytext)
-                    msg.append(yytext);
+                msg.append(yytext);
                 msg.append("' is already defined");
-                reportError(ERR_ID_REDEFINE,msg.str(), lineno, column, pos);
-                return;
+                syntaxErrorCompareExpected(msg, yytext, expected, expectedLength);
             }
+            reportError(ERR_ID_REDEFINE,msg.str(), lineno, column, pos);
+            return;
         }
-        // fall into...
     }
-    
-    default:
-        msg.append(s);
-        if (yytext && *yytext)
-        { 
-            if (yytext[0]=='\n')
-                msg.append(" near the end of the line");
-            else
-                msg.append(" near \"").append(yytext).append('\"');
-        }
 
+    msg.append(s);
+    if (yytext && *yytext)
+    {
+        if (yytext[0]=='\n')
+            msg.append(" near the end of the line");
+        else
+            msg.append(" near \"").append(yytext).append('\"');
+    }
+
+    if (isMajority(expected, expectedLength, isIdentifier))
+        msg.append(" : expected identifier");
+    else if (isMajority(expected, expectedLength, isExpression))
+        msg.append(" : expected expression");
+    else if (isMajority(expected, expectedLength, isFunction))
+        msg.append(" : expected function");
+    else if (isMajority(expected, expectedLength, isType))
+        msg.append(" : expected type-name");
+    else
+    {
+        //Report a filtered, limited, & deduped expected list
         bool first = true;
-        for (int i = 0; expected[i] ; i++)
+        StringArray expectedToReport;
+        for (unsigned i = 0; i < MAX_EXPECTED_REPORTED; i++)
+        {
+            /*The collection of the expected tokens has no specific structure in that they are ordered
+             *in no really useful mannar. It thus makes sense to not simply report the first N expected
+             *but instead, report N expected tokens equally distributed over the expected set.
+             */
+            unsigned indx = expectedLength <= MAX_EXPECTED_REPORTED ? i : i*expectedLength/MAX_EXPECTED_REPORTED - 1;
+            StringBuffer temp;
+            getSensibleTokenText(temp, expected[indx]);
+            if (!expectedToReport.appendUniq(temp.str()))
+            {
+                //Search for a unique expected token to report
+                for (unsigned j = 0; j < expectedLength; j++)
+                {
+                    StringBuffer temp;
+                    getSensibleTokenText(temp, expected[indx]);
+                    if (expectedToReport.appendUniq(temp.str()))
+                        break;
```

Comment:
The checking for unique strings here and else-where is not something I particularly like either, but I wouldn't reject just because of that (because this code is not time critical).